src/cpp_generator.cpp

```diff
@@ -207,9 +208,22 @@ cpp_generator::function cpp_generator::generate_module(const module& m,
     });
     f.set_name(name).set_types(m).set_body(
         m, [&](instruction_ref ins, const auto& names) -> std::string {
-            if(ins->name() == "@literal")
-                return shape::cpp_type(ins->get_shape().type()) + "(" +
-                       ins->get_literal().to_string() + ")";
+            if(ins->name() == "@literal"){
+                auto string_literal = ins->get_literal().to_string();
+
+                if (string_literal == "inf") {
+                    string_literal =  std::numeric_limits<decltype(shape::cpp_type(ins->get_shape().type()))>::max() + "+ 1";
```

Comment:
This is wrong. The type of `decltype(shape::cpp_type(ins->get_shape().type()))` is `shape::type_t` not a floating point. And this probably points to a string in random memory.

You need to use a visitor here:

```cpp
std::string string_literal;
auto cpp_type = shape::cpp_type(ins->get_shape().type());
ins->get_literal().visit([&](auto v) {
    assert(v.size() == 1); 
    auto x = v.front();
    if (std::isinf(x)) {
        string_literal = "std::numeric_limits<" + cpp_type + ">::infinity()";
    } else if (std::isnan(x)) {
        string_literal = "std::numeric_limits<" + cpp_type + ">::quietNaN()";
    } else {
        string_literal = std::to_string(x);
    }
});
```

Although this wont work with hiprtc.

Reply:
Right. Totally missed the visit part. I think we can get around using std::numeric limits by just performing the floating point division between 1.0/0.0 with proper sign here since that will get evaluated at compile as INF correctly when the new code is generated. This should work for both float and double, since and inf promoted from float to double are treated the same

Reply:
Resolved since we're using builtins__*** now